### PR TITLE
ruler: Support OAuth2 and proxies in Alertmanager client

### PR DESCRIPTION
**Pull request by alexweav**
#### What this PR does

The AM client already supports several TLS options + basic auth for talking to Alertmanager.
Extend this to also support OAuth2 Client Credentials Workflow, plus a URL for talking to the Alertmanager through a proxy.

These are all supported in upstream Prometheus's Alertmanager client: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#alertmanager_config
via `http_config: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#alertmanager_config

Supports HTTP, HTTP with CONNECT, or SOCKS5 proxies per https://pkg.go.dev/net/http#Transport. Under the hood, `prometheus/common` doesn't implement anything special for proxies and just wires the configuration up to `http.Transport`'s proxy ([1](https://github.com/prometheus/common/blob/7ed45232224c2722ad677589452579095a42e4fe/config/http_config.go#L1478), [2](https://github.com/prometheus/common/blob/7ed45232224c2722ad677589452579095a42e4fe/config/http_config.go#L592))

#### Which issue(s) this PR fixes or relates to

n/a

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Comment by alexweav** [comment]
Still in draft state as I finalize testing.

File: docs/sources/mimir/configure/configuration-parameters/index.md [hunk 1]
```diff
@@ -1946,6 +1946,30 @@ alertmanager_client:
...
+    # OAuth2 client ID. Enables the use of OAuth2 for authenticating with
+    # Alertmanager.
+    # CLI flag: -ruler.alertmanager-client.oauth.client_id
+    [client_id: <string> | default = ""]
+
+    # OAuth2 client secret.
+    # CLI flag: -ruler.alertmanager-client.oauth.client_secret
+    [client_secret: <string> | default = ""]
+
+    # Endpoint used to fetch access token from.
```
Comment (tacole02) [inline docs/sources/mimir/configure/configuration-parameters/index.md:1959]:
```suggestion
    # Endpoint used to fetch the access token.
```

File: docs/sources/mimir/configure/configuration-parameters/index.md [hunk 2]
```diff
@@ -1946,6 +1946,30 @@ alertmanager_client:
...
+    # CLI flag: -ruler.alertmanager-client.oauth.token_url
+    [token_url: <string> | default = ""]
+
+    # Optional scopes to include with the token request.
+    # CLI flag: -ruler.alertmanager-client.oauth.scopes
+    [scopes: <list of strings> | default = []]
+
+  # (advanced) Optional HTTP, HTTPS via CONNECT, or SOCKS5 proxy URL to route
+  # requests through. Applies to all requests, including infra like oauth token
```
Comment (tacole02) [inline docs/sources/mimir/configure/configuration-parameters/index.md:1968]: `infra` = `infrastructure`?

Comment (tacole02) [inline docs/sources/mimir/configure/configuration-parameters/index.md:1968]:
```suggestion
  # requests through. Applies to all requests, including infra like OAuth token
```

**Review by tacole02 — APPROVED** [review]
Looks good!

File: docs/sources/mimir/configure/configuration-parameters/index.md (hunk 2 above)
Comment (alexweav) [reply docs/sources/mimir/configure/configuration-parameters/index.md:1968]: Good catch, that wasn't the greatest wording.
How about: `Applies to all requests, including auxiliary traffic such as OAuth token requests.`
The intent meaning: "this applies all requests related to the client, not just the obvious Alertmanager ones"

Comment (tacole02) [reply docs/sources/mimir/configure/configuration-parameters/index.md]: That sounds great! Just add another comma after traffic:

`Applies to all requests, including auxiliary traffic, such as OAuth token requests.`

File: pkg/ruler/notifier.go [hunk 3]
```diff
@@ -27,18 +28,43 @@ import (
...
 	cfg.BasicAuth.RegisterFlagsWithPrefix("ruler.alertmanager-client.", f)
+	cfg.OAuth2.RegisterFlagsWithPrefix("ruler.alertmanager-client.oauth.", f)
+	f.StringVar(&cfg.ProxyURL, "ruler.alertmanager-client.proxy-url", "", "Optional HTTP, HTTPS via CONNECT, or SOCKS5 proxy URL to route requests through. Applies to all requests, including auxiliary traffic, such as OAuth token requests.")
+}
+
+type OAuth2Config struct {
+	ClientID     string              `yaml:"client_id"`
+	ClientSecret flagext.Secret      `yaml:"client_secret"`
+	TokenURL     string              `yaml:"token_url"`
+	Scopes       flagext.StringSlice `yaml:"scopes,omitempty"`
```
Comment (alexweav) [inline pkg/ruler/notifier.go:56]: Note: Moving this to StringSliceCSV. Nearly every other field that behaves this way across the project uses the CSV version, the regular version looks like it doesn't work as cleanly.

**Comment by francoposa** [comment]
per discussion we said we would leave adding the support for specifying token audience via client credentials config `EndpointParams` for a subsequent PR.

**Pull request merged**